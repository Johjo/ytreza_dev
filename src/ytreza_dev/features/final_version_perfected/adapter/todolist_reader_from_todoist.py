from toml.encoder import TomlPreserveInlineDictEncoder

from ytreza_dev.features.final_version_perfected.port.todolist_reader import TodolistReaderPort
from ytreza_dev.features.final_version_perfected.types import ExternalTask, ExternalProject
from ytreza_dev.shared.env_reader import EnvReaderPort
from ytreza_dev.shared.todoist.todoist_api import TodoistAPI, TodoistTask, TodoistProject


class TodolistReaderFromTodoist(TodolistReaderPort):
    def __init__(self, env_reader: EnvReaderPort):
        self._env_reader = env_reader

    def all_tasks(self) -> list[ExternalTask]:
        todoist = TodoistAPI(self._env_reader.read("TODOIST_API_TOKEN"))
        tasks = todoist.get_all_tasks()
        return [self._to_external_task(task) for task in tasks]

    def _to_external_task(self, task: TodoistTask):
        return ExternalTask(name=task.name, url=task.url, id=task.id, project=self._to_external_project(task.project))

    def _to_external_project(self, project: TodoistProject) -> ExternalProject:
        return ExternalProject(name=project.name, key=project.id)
