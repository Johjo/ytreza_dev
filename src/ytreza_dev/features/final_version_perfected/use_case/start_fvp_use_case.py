from expression import Nothing

from ytreza_dev.features.final_version_perfected.port.task_information_repository import TaskInformationRepositoryPort, \
    TaskInformation
from ytreza_dev.features.final_version_perfected.port.task_repository import FvpRepositoryPort
from ytreza_dev.features.final_version_perfected.port.todolist_reader import TodolistReaderPort
from ytreza_dev.features.final_version_perfected.types import TaskNew, ExternalTask, ExternalProject, Project


class StartFvpUseCase:
    def __init__(self, todolist_reader: TodolistReaderPort, fvp_repository: FvpRepositoryPort, task_information_repository: TaskInformationRepositoryPort) -> None:
        self._task_information_repository = task_information_repository
        self._todolist_reader = todolist_reader
        self._fvp_repository = fvp_repository

    def execute(self) -> None:

        all_external_tasks : list[ExternalTask] = self._todolist_reader.all_tasks()
        self._fvp_repository.save([self._to_task_new(task) for task in all_external_tasks])
        self._task_information_repository.save([self._to_task_information(task) for task in all_external_tasks])

    @staticmethod
    def _to_task_information(external_task: ExternalTask) -> TaskInformation:
        return TaskInformation(
            key=external_task.id,
            title=external_task.name,
            project=Project(key=external_task.project.key, name=external_task.project.name),
            url=external_task.url,
            due_date=Nothing
        )

    @staticmethod
    def _to_task_new(task: ExternalTask) -> TaskNew :
        return TaskNew(id=task.id)

    @staticmethod
    def _to_project(project: ExternalProject) -> Project:
        return Project(key=project.key, name=project.name)
