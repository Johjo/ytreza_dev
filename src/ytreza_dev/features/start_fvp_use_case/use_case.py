from abc import ABCMeta, abstractmethod
from dataclasses import dataclass

from ytreza_dev.shared.final_version_perfected.types import TaskBase, TaskNew


@dataclass
class OldTask:
    name: str
    url: str


class TaskRepository(metaclass=ABCMeta):
    @abstractmethod
    def save(self, tasks: list[TaskBase]) -> None:
        pass

@dataclass
class ExternalTask:
    name: str
    url: str


class TodolistReader(metaclass=ABCMeta):
    @abstractmethod
    def all_tasks(self) -> list[ExternalTask]:
        pass


class StartFvpUseCase:
    def __init__(self, todolist_reader: TodolistReader, task_repository: TaskRepository) -> None:
        self._todolist_reader = todolist_reader
        self._task_repository = task_repository

    def execute(self) -> None:
        self._task_repository.save([TaskNew(title=task.name, url=task.url) for task in self._todolist_reader.all_tasks()])
