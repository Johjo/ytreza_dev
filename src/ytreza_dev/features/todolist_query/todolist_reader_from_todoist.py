import os
import requests
from dotenv import load_dotenv

from ytreza_dev.features.todolist_query.todolist_query import TodolistReaderPort, Task

load_dotenv()


class TodolistReaderFromTodoist(TodolistReaderPort):
    def __init__(self):
        self.api_token = os.getenv("TODOIST_API_TOKEN")
        if not self.api_token:
            raise ValueError("TODOIST_API_TOKEN environment variable not set.")
        self.api_url = "https://api.todoist.com/rest/v2"

    def all_tasks(self) -> list[Task]:
        headers = {
            "Authorization": f"Bearer {self.api_token}",
        }
        response = requests.get(f"{self.api_url}/tasks", headers=headers)
        response.raise_for_status()  # Raise an exception for bad status codes
        tasks_data = response.json()
        return [Task(name=task_data["content"], url=task_data["url"]) for task_data in tasks_data]

