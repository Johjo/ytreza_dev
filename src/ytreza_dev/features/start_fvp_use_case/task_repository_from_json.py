import json
from dataclasses import asdict
from pathlib import Path

from ytreza_dev.features.start_fvp_use_case.use_case import TaskRepository, OldTask
from ytreza_dev.shared.final_version_perfected.types import TaskBase


class TaskRepositoryFromJson(TaskRepository):
    def all_tasks(self) -> list[TaskBase]:
        raise NotImplementedError()

    def __init__(self, file_path: Path) -> None:
        self._file_path = file_path

    def save(self, tasks: list[TaskBase]) -> None:
        tasks_dict = [self._to_dict(task) for task in tasks]
        self._file_path.write_text(json.dumps(tasks_dict, indent=4))

    @staticmethod
    def _to_dict(task: TaskBase):
        d = asdict(task)
        d.update({"status": "new"})
        return d
