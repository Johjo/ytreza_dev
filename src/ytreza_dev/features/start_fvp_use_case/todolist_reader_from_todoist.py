from ytreza_dev.features.start_fvp_use_case.use_case import TodolistReader, OldTask, ExternalTask
from ytreza_dev.shared.env_reader import EnvReaderPort, EnvReaderFromEnv
from ytreza_dev.shared.todoist.todoist_api import TodoistAPI


class TodolistReaderFromTodoist(TodolistReader):
    def __init__(self, env_reader: EnvReaderPort):
        self._env_reader = env_reader

    def all_tasks(self) -> list[ExternalTask]:
        todoist = TodoistAPI(self._env_reader.read("TODOIST_API_TOKEN"))
        tasks = todoist.get_all_tasks()
        return [ExternalTask(name=task.name, url=task.url) for task in tasks]
