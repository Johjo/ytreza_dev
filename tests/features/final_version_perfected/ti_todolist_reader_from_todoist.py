import os

from approvaltests import verify # type: ignore
from dotenv import load_dotenv

from ytreza_dev.features.start_fvp_use_case.todolist_reader_from_todoist import TodolistReaderFromTodoist


def test_all_tasks() -> None:
    load_dotenv(dotenv_path=".env.test")
    api_token = os.getenv("TODOIST_API_TOKEN")
    assert api_token, "TODOIST_API_TOKEN not found in .env.test"

    todoist_api = TodolistReaderFromTodoist()
    tasks = todoist_api.all_tasks()

    verify("\n".join([str(task) for task in tasks]), encoding="utf-8")
