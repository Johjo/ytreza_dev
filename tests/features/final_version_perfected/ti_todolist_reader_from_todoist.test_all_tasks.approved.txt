Task(name='T√©l√©chargez des [applications et plugins gratuits suppl√©mentaires](https://todoist.com/fr/downloads?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=inbox) pour üíª,‚åöÔ∏è,üñ•,üì± et üìß', url='https://app.todoist.com/app/task/9388991121')
Task(name='Connecter mon calendrier professionnel √† Todoist', url='https://app.todoist.com/app/task/9388991491')
Task(name='R√©pondez au [questionnaire sur votre m√©thode de productivit√©](https://todoist.com/productivity-methods?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=inbox)', url='https://app.todoist.com/app/task/9388991122')
Task(name='Passer en revue ma journ√©e et planifier', url='https://app.todoist.com/app/task/9388991492')
Task(name='Ajouter mes routines _professionnelles_', url='https://app.todoist.com/app/task/9388991493')
Task(name='[Faire une pause r√©guli√®re](https://blog.doist.com/end-work-day/?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=work) üåñ', url='https://app.todoist.com/app/task/9388991494')
Task(name='[Les int√©grations de Todoist qui boosteront votre productivit√©](https://blog.doist.com/todoist-integrations-productivity/?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=work)', url='https://app.todoist.com/app/task/9388991495')
Task(name='Explorer les [mod√®les professionnels de Todoist](https://todoist.com/fr/templates/category/work?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=work)', url='https://app.todoist.com/app/task/9388991496')
